babel-plugin-transform-jsx/lib/index.js

```diff
@@ -154,21 +58,139 @@ export default function ({ types: t }) {
   }
 
   /* ==========================================================================
+   * Visitors
+   * ======================================================================= */
+
+  const visitJSXElement = (path, state) => {
+
+    const {
+      opts: {
+        useVariables = false
+      }
+    } = state
+
+    /* ==========================================================================
+     * Node Transformers
+     * ======================================================================= */
+
+    const JSXIdentifier = node => t.stringLiteral(node.name)
+
+    const JSXNamespacedName = node => t.stringLiteral(`${node.namespace.name}:${node.name.name}`)
+
+    const JSXMemberExpression = transformOnType({
+      JSXIdentifier: node => t.identifier(node.name),
+      JSXMemberExpression: node => (
+        t.memberExpression(
+          JSXMemberExpression(node.object),
+          JSXMemberExpression(node.property)
+        )
+      )
+    })
+
+    const JSXElementName = transformOnType({
+      JSXIdentifier: useVariables
+        ? node => path.scope.hasBinding(node.name) ? t.identifier(node.name) : JSXIdentifier(node)
```

Comment:
My initial idea was to use a regular expression option and if it passed a variable reference would be used with the first capture grouped (if specified as the variable name) otherwise a string would be used.

For example `^[A-Z]` would make any React-styled component with the first letter as uppercase a variable, or `^&(.*)` which would make anything starting with an ampersand a variable (the following characters would be the name).

Could you make a quick argument for doing it this way? I think it may be a bit confusing. In addition sometimes when writing forms I have an `input` variable in scope which may be confusing. I think I'm cool with it, I just want a justification for it.